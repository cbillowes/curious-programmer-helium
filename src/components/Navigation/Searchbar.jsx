import React, { Component } from "react"
import "./Searchbar.scss"

class SearchBar extends Component {
  constructor(props) {
    super(props)
    this.state = {
      query: "",
      loaded: false,
      results: []
    }
    this.searchQuery = React.createRef()
  }

  generateResults = () => {
    if (this.state.error) {
      return (
        <li key="searchbar-empty" className="empty">
          Sorry, but something strange is going on. Search does not want to
          behave appropriately. Perhaps try again a little later. Drop me a line
          if it persists.
        </li>
      )
    }

    if (this.state.loaded && this.state.query !== "") {
      let numberOfResults = this.state.results.length
      if (numberOfResults === 0) {
        return (
          <li key="searchbar-empty" className="empty">
            Nothing came up. Go on, give it another shot.
          </li>
        )
      }

      let results = []
      for (let i = 0; i < numberOfResults; i++) {
        let result = this.state.results[i]
        let title = result.title
          .toString()
          .replace(/\| Curious Programmer/gi, "")
        let url = result.url
          .toString()
          .replace(/https:\/\/curiousprogrammer.io/gi, "")
        results.push(
          <li key={`searchbar-${i}`} onClick={this.props.toggle.bind(this)}>
            <a href={url}>{title}</a>
          </li>
        )
      }
      return results
    } else {
      return (
        <li key="searchbar-empty" className="empty">
          What do you feel like reading today? Search as you type, also, enter
          works too.
        </li>
      )
    }
  }

  query = (e) => {
    let query = this.searchQuery.current.value
    if (query.length > 4 || e.keyCode === 13) {
      if (query !== this.state.query) {
        this.search(query)
      }
    }

    if (query.length === 0) {
      this.setState({ query: "" })
    }
  }

  search = (query) => {
    fetch(
      `https://curiousprogrammer.tk/solr/oxygen/select?q=text:${encodeURIComponent(
        query
      )}&fl=url,title&rows=20&wt=json`
    )
      .then((res) => res.json())
      .then(
        (result) => {
          let results = []
          if (result.response) {
            results = result.response.docs
          }
          this.setState({
            query: query,
            loaded: true,
            results: results
          })
        },
        (error) => {
          this.setState({
            loaded: true,
            error: error
          })
        }
      )
  }

  render() {
    return (
      <div className={`searchbar ${this.props.active ? "active" : ""}`}>
        <input
          ref={this.searchQuery}
          className="search-query"
          type="text"
          onKeyUp={this.query.bind(this)}
        />
        <div className="results">
          <ul>{this.generateResults()}</ul>
        </div>
      </div>
    )
  }
}

export default SearchBar
