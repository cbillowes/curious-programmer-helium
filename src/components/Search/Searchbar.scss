@import "../../layout/variables.scss";

.searchbar {
    display: none;
    background-color: $background-darker;
    position: fixed;
    left: 0;
    bottom: 0;
    right: 0;
    top: 4em;
    z-index: 1;
    padding: 1em;

    &.active {
        display: block;
    }

    .search-query {
        display: block;
        width: 96%;
        padding: 0.5em;
        margin: 1.5em 2% 0.5em 2%;
        font-family: $font-family-sans-serif;
        background-color: $background-lighter;
        border: solid 1px $background-lighter;
        border-radius: $border-radius;
        color: $foreground;
        font-size: 1.25em;
    }

    .error {
        .message {
            color: $accent;
            font-weight: $font-family-sans-serif-heavy;
        }

        .contact-me {
            a {
                display: inline-block;
            }
        }
    }

    .results {
        height: 100vh;
        overflow-y: auto;
        font-size: $font-size * 1.15;

        ul {
            padding-left: 2%;

            li {
                list-style: none;
                padding-bottom: 0.5em;

                a {
                    display: block;
                }
            }
        }
    }
}
