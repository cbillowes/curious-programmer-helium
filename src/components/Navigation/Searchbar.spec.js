import React from "react"
import renderer from "react-test-renderer"
import { mount } from "enzyme"
import Searchbar from "./Searchbar"

describe("a search bar component", () => {
  let element = <Searchbar toggle={() => {}} active="true" />
  let component
  
  beforeAll(() => {
    component = mount(element)    
  })

  it("should render", () => {
    const tree = renderer.create(element).toJSON()
    expect(tree).toMatchSnapshot()
  })
})