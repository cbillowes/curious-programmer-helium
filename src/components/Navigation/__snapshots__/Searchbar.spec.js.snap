// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a search bar component should render 1`] = `
<div
  className="searchbar active"
>
  <input
    className="search-query"
    onKeyUp={[Function]}
    type="text"
  />
  <div
    className="results"
  >
    <ul>
      <li
        className="empty"
      >
        What do you feel like reading today? Search as you type, also, enter works too.
      </li>
    </ul>
  </div>
</div>
`;
